quickwit-search/src/fetch_docs.rs

```diff
@@ -144,30 +163,140 @@ async fn get_searcher_for_split_without_cache(
     Ok(reader)
 }
 
+/// A struct for holding a fetched document's content and highlights.
+struct DocBundle {
+    doc_json: String,
+    highlights_json: Option<String>,
+}
+
 /// Fetching docs from a specific split.
 #[tracing::instrument(skip(global_doc_addrs, index_storage, split))]
 #[allow(clippy::needless_lifetimes)]
 async fn fetch_docs_in_split(
     mut global_doc_addrs: Vec<GlobalDocAddress>,
     index_storage: Arc<dyn Storage>,
     split: &SplitIdAndFooterOffsets,
-) -> anyhow::Result<Vec<(GlobalDocAddress, String)>> {
+    doc_mapper: Arc<dyn DocMapper>,
+    search_request: &SearchRequest,
+) -> anyhow::Result<Vec<(GlobalDocAddress, DocBundle)>> {
     global_doc_addrs.sort_by_key(|doc| doc.doc_addr);
 
     let index_reader = get_searcher_for_split_without_cache(index_storage, split).await?;
-    let searcher = Arc::new(index_reader.searcher());
+    let searcher = index_reader.searcher();
+    let snippet_generators =
```

Comment:
Maybe we can have a struct like `MultiFieldSnippetGenerator`

Reply:
Apart from putting code related to snippet generation inside, one pros is that we will have a variable `Option< FieldsSnippetGenerator>` that show clearly that we use the generator only when there are snippet fields.